kasa/tapo/tapodevice.py

```diff
@@ -0,0 +1,160 @@
+        *,
+        port: Optional[int] = None,
+        credentials: Optional[Credentials] = None,
+        timeout: Optional[int] = None,
+    ) -> None:
+        super().__init__(host, port=port, credentials=credentials, timeout=timeout)
+        self._state_information: Dict[str, Any] = {}
+        self._discovery_info: Optional[Dict[str, Any]] = None
+        self.protocol = TPLinkAes(host, credentials=credentials, timeout=timeout)
+
+    async def update(self, update_children: bool = True):
+        """Update the device."""
+        if self.credentials is None or self.credentials.username is None:
+            raise AuthenticationException("Tapo plug requires authentication.")
+
+        self._info = await self.protocol.query("get_device_info")
+        self._usage = await self.protocol.query("get_device_usage")
+        self._time = await self.protocol.query("get_device_time")
+
+        self._last_update = self._data = {
+            "info": self._info,
+            "usage": self._usage,
+            "time": self._time,
+        }
+
+        _LOGGER.debug("Got an update: %s", self._data)
+
+    @property
+    def sys_info(self) -> Dict[str, Any]:
+        """Returns the device info."""
+        return self._info
+
+    @property
+    def model(self) -> str:
+        """Returns the device model."""
+        return str(self._info.get("model"))
+
+    @property
+    def alias(self) -> str:
+        """Returns the device alias or nickname."""
+        return base64.b64decode(str(self._info.get("nickname"))).decode()
+
+    @property
+    def time(self) -> datetime:
+        """Return the time."""
+        return datetime.fromtimestamp(
+            cast(float, self._time.get("timestamp")),
+            tz=timezone(timedelta(minutes=cast(float, self._time.get("time_diff")))),
+        )
+
+    @property
+    def timezone(self) -> Dict:
+        """Return the timezone and time_difference."""
+        return {
+            "timezone": timezone(
+                timedelta(minutes=cast(float, self._time.get("time_diff")))
+            )
+        }
+
+    @property
+    def hw_info(self) -> Dict:
+        """Return hardware info for the device."""
+        return {
+            "sw_ver": self._info.get("fw_ver"),
+            "hw_ver": self._info.get("hw_ver"),
+            "mac": self._info.get("mac"),
+            "type": self._info.get("type"),
+            "hwId": self._info.get("device_id"),
+            "dev_name": self.alias,
+            "oemId": self._info.get("oem_id"),
+        }
+
+    @property
+    def location(self) -> Dict:
+        """Return the device location."""
+        loc = {
+            "latitude": cast(float, self._info.get("latitude")) / 10_000,
+            "longitude": cast(float, self._info.get("longitude")) / 10_000,
+        }
+        return loc
+
+    @property
+    def rssi(self) -> Optional[int]:
+        """Return the rssi."""
+        rssi = self._info.get("rssi")
+        return int(rssi) if rssi else None
```

Comment:
Same here (and below also) regarding to casting. I'll skip adding separate comments on this topic but if changes are needed please adapt other pieces of code as needed :-)

Reply:
Resolving this as we have two open cast convos above